### PR TITLE
Remove usage of AstNode in checks using array designators

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/ArrayDesignatorOnVariableCheck.java [hunk 1]
```diff
@@ -42,18 +42,29 @@
...
-    if (node.hasDirectChildren(TreeFactory.WRAPPER_AST_NODE) || (node.is(Kind.VARIABLE) && ((VariableTreeImpl) node).dims() > 0)) {
-      getContext().createLineViolation(this, "Move the array designator from the variable to the type.", node);
+  public void visitNode(Tree tree) {
+    VariableTree variableTree = (VariableTree) tree;
+    TypeTree type = variableTree.type();
+    int identifierColumn = variableTree.simpleName().identifierToken().column();
+    while (type.is(Tree.Kind.ARRAY_TYPE)) {
+      ArrayTypeTree arrayTypeTree = (ArrayTypeTree) type;
+      SyntaxToken arrayDesignatorToken = arrayTypeTree.ellipsisToken();
+      if (arrayDesignatorToken == null) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/ArrayDesignatorOnVariableCheck.java]: CRITICAL Redundant nullcheck of arrayDesignatorToken, which is known to be non-null in org.sonar.java.checks.ArrayDesignatorOnVariableCheck.visitNode(Tree) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ARCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE)

File: java-squid/src/main/java/org/sonar/java/model/declaration/AnnotatedDimension.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+  public SyntaxToken closeBracketToken() {
+    return closeBracketToken;
+  }
+
+  public SyntaxToken ellipsisToken() {
+    return ellipsisToken;
+  }
+
+  public List<AstNode> getAstNodes() {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/model/declaration/AnnotatedDimension.java]: MAJOR org/sonar/java/model/declaration/AnnotatedDimension must not use com/sonar/sslr/api/AstNode [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AAstNode_SSLR_class_should_not_be_used)

File: java-squid/src/main/java/org/sonar/java/model/JavaTree.java [hunk 3]
```diff
@@ -605,5 +629,25 @@ public void accept(TreeVisitor visitor) {
     public Iterator<Tree> childrenIterator() {
       return Iterators.<Tree>singletonIterator(type);
     }
+
+    @Override
+    public List<AnnotationTree> annotations() {
+      return (List<AnnotationTree>) annotations;
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/model/JavaTree.java]: MINOR Remove this unnecessary cast to "List". [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1905)

File: java-squid/src/main/java/org/sonar/java/model/expression/MemberSelectExpressionTreeImpl.java [hunk 4]
```diff
@@ -67,18 +69,17 @@ public MemberSelectExpressionTreeImpl(ExpressionTree expression, IdentifierTree
...
 
-    this.dims = -1;
+    this.dims = ImmutableList.<AnnotatedDimension>of();
     this.expression = Preconditions.checkNotNull(expression);
     this.identifier = Preconditions.checkNotNull(identifier);
   }
 
   public MemberSelectExpressionTreeImpl completeWithExpression(ExpressionTree expression) {
-    Preconditions.checkState(dims >= 0 && this.expression == null);
+    Preconditions.checkState(dims.size() >= 0 && this.expression == null);
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/model/expression/MemberSelectExpressionTreeImpl.java]: CRITICAL Bad comparison of nonnegative value with 0 in org.sonar.java.model.expression.MemberSelectExpressionTreeImpl.completeWithExpression(ExpressionTree) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3AINT_BAD_COMPARISON_WITH_NONNEGATIVE_VALUE)

File: java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java [hunk 5]
```diff
@@ -2365,26 +2340,82 @@ private void add(Object o) {
...
+      ImmutableList.Builder<AnnotatedDimension> dimensionsBuilder = ImmutableList.builder();
+      for (Tuple<AstNode, AstNode> dim : dims.get()) {
+        dimensionsBuilder.add(getDimension(dim));
+      }
+      dimensionsWithTokens = dimensionsBuilder.build();
+    }
+    return dimensionsWithTokens;
+  }
+
+  private AnnotatedDimension getDimension(Tuple<AstNode, AstNode> dim) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java]: MAJOR Make "getDimension" a "static" method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2325)

File: java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java [hunk 6]
```diff
@@ -2365,26 +2340,82 @@ private void add(Object o) {
...
+      ImmutableList.Builder<AnnotatedDimension> dimensionsBuilder = ImmutableList.builder();
+      for (Tuple<Optional<List<AnnotationTreeImpl>>, Tuple<AstNode, AstNode>> dim : dims.get()) {
+        dimensionsBuilder.add(getAnnotatedDimension(dim));
+      }
+      dimensionsWithTokens = dimensionsBuilder.build();
+    }
+    return dimensionsWithTokens;
+  }
+
+  private AnnotatedDimension getAnnotatedDimension(Tuple<Optional<List<AnnotationTreeImpl>>, Tuple<AstNode, AstNode>> dim) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java]: MAJOR Make "getAnnotatedDimension" a "static" method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 new issues:
* 2 critical
* 3 major
* 1 minor

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 new issues:
* 2 critical
* 3 major
* 1 minor

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 1 critical
* 1 major

File: java-checks/src/main/java/org/sonar/java/checks/ArrayDesignatorOnVariableCheck.java (hunk 1 above)
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/ArrayDesignatorOnVariableCheck.java]: ellipsisToken() is annotated with @Nullable.

File: java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java [hunk 7]
```diff
@@ -2093,20 +2081,15 @@ public ExpressionTree completeCreatorSelector(AstNode dotTokenAstNode, Expressio
...
+    if (nestedAnnotations != null) {
+      children.add(nestedAnnotations);
     }
-
     children.add(dotTokenAstNode);
     children.add(identifier);
 
-    return new MemberSelectExpressionTreeImpl(dimensions.isPresent() ? dimensions.get().size() : 0, identifier,
-      children);
+    return new MemberSelectExpressionTreeImpl(nestedAnnotations, identifier, children);
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java]: CRITICAL Null passed for non-null parameter of new org.sonar.java.model.expression.MemberSelectExpressionTreeImpl(JavaTree$ArrayTypeTreeImpl, IdentifierTreeImpl, List) in org.sonar.java.ast.parser.TreeFactory.newDotClassSelector(Optional, AstNode, AstNode) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_NULL_PARAM_DEREF)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 critical

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 new issue:
* 1 critical

File: java-checks/src/main/java/org/sonar/java/checks/ArrayDesignatorOnVariableCheck.java [hunk 8]
```diff
@@ -42,19 +42,35 @@
...
-    if (node.hasDirectChildren(TreeFactory.WRAPPER_AST_NODE) || (node.is(Kind.VARIABLE) && ((VariableTreeImpl) node).dims() > 0)) {
-      getContext().createLineViolation(this, "Move the array designator from the variable to the type.", node);
+  public void visitNode(Tree tree) {
+    VariableTree variableTree = (VariableTree) tree;
+    TypeTree type = variableTree.type();
+    SyntaxToken identifierToken = variableTree.simpleName().identifierToken();
+    while (type.is(Tree.Kind.ARRAY_TYPE)) {
+      ArrayTypeTree arrayTypeTree = (ArrayTypeTree) type;
+      SyntaxToken arrayDesignatorToken = arrayTypeTree.ellipsisToken();
+      if (arrayDesignatorToken == null) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/ArrayDesignatorOnVariableCheck.java]: CRITICAL Redundant nullcheck of arrayDesignatorToken, which is known to be non-null in org.sonar.java.checks.ArrayDesignatorOnVariableCheck.visitNode(Tree) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ARCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 new issue:
* 1 critical

File: java-checks/src/main/java/org/sonar/java/checks/ArrayDesignatorOnVariableCheck.java (hunk 8 above)
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/ArrayDesignatorOnVariableCheck.java]: `ellipsisToken()` is annotated with `@Nullable`.

Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/ArrayDesignatorOnVariableCheck.java]: Why not with `@CheckForNull` ?

**Comment by benzonico** [comment]
Please squash two first commits as discussed, otherwise, LGTM

**Pull request merged**